### DISCUSSION TITLE
Modular implicits syntax

### DISCUSSION THREAD

**Discussion by samsa1** [ocaml/ocaml#13274] [Features]
# Syntax questions around explicit and implicit module applications

I am working on [Modular implicits](https://arxiv.org/pdf/1512.01895), an experimental OCaml feature that is intended as a _typeclass-like_ mechanism. The idea behind it is to have an implicit argument mechanism with module arguments. There remain important unsolved design questions around modular implicits and we don't expect them to be integrated in the language before at least a few years.

It has been suggested in the past to integrate _modular explicits_ first, which permits writing module dependent functions without providing any implicit inference or elaboration. I believe that this is a nice idea, and am preparing a new PR to propose modular explicits for integration in the next few weeks/months.

There are thorny questions around the syntax we want for those features. We plan to propose only modular *explicits* at first, but we want to pick a syntax that integrates well with the language today, but will also integrate well with modular *implicits* proposed in the future.

In this post we propose to discuss the syntax for both (and related OCaml features) to have a better view of the (syntactic) design space. It presents 4 different syntaxes that I have heard about for _modular explicits/implicits_ (the 4th being my favorite but the most complicated).

#### Proposal by Matthew Ryan

The first [PR for modular explicits by M. Ryan](https://github.com/ocaml/ocaml/pull/9187) proposed to reuse the syntax of modular implicits presented above for modular explicits: `let f {M : S} x ...` for explicit module abstractions, and `f {Int} foo` for explicit module applications.

The idea is that if we add modular implicits later, then all examples can be re-intepreted as modular implicits, and now providing explicit applications is optional -- explicit applications `f {Int} foo` are still supported, but users can write just `f foo` and the module arguments gets (implicitly) elaborated.

#### Problem: split between first-class modules and dependent module arguments

The main issue with this proposal is the split between first-class modules and dependent module arguments. The two features are very related (a function can take a module as parameter), but they have different syntaxes and different typing rules:

- `let f (module M : S) = body` can receive any first-class module as argument, however `M` cannot appear in the type of `body`
- `let f {M : S} = body` must receive a module expression at application, but `M` can appear in the type of `body`

How will non-expert users know which one to use ?

This problem was spotted by @EduardoRFS (see [github](https://github.com/ocaml/ocaml/pull/9187#issuecomment-776130420) or [discuss](https://discuss.ocaml.org/t/fcm-and-modular-implicits-type-system/7085)).

- On one hand we have first-class modules that are known by the community, they are a regular type of value but cannot be made dependent.

- On the other hand there are modular explicits with a completely different syntax that will be brand new. It is not a value thus we cannot have a function that returns a module using this mechanism, but functions can be made dependent.

Ideally we could use a syntax for modular explicits that coincides with first-class modules, at least in the cases where both can be used, and yet integrates gracefully with modular implicits.

In order to think and discuss about this problem we will present and give arguments for and against various syntaxes for modular implicits and explicits.

### 1 - The initial proposal for modular implicits

In 2014 Leo White, Frédéric Bour and Jeremy Yallop proposed a syntax with brackets for implicit arguments :

```ocaml
module type Show = sig
    type t
    val show : t -> string
end

let show {S : Show} x = S.show x
(* val show : {S : Show} -> S.t -> string *)

let () =
    print_string (show 3);
    print_string (show {Variant_Int_Show} 3)
```

Pros of this syntax :
- This syntax allows a compact syntax for implicit arguments that might be frequent.
- There is a clear distinction between implicit arguments and regular arguments.

Cons of this syntax :
- We have this unergonomic split between first-class modules and modular implicits.
- At the call-site the user decides to provide an explicit module argument or not, but at the function-declaration site it is not possible to force all call-sites to explicitly provide an argument, which is sometimes desired for readability or for functions that cannot be inferred implicitly.
- If we have 3 implicits argument, and we want to instantiate the third by hand, we need to instantiate the other two.

### 2 - New proposal : Modular implicits as optional modular explicits

The idea here would be to use the syntax presented above for modular explicits as done by [Matthew Ryan](https://github.com/ocaml/ocaml/pull/9187), but with labels to module arguments.

```ocaml
(** always applied explicitly *)
let f1 {M : S} = body1

(** always applied explicitly but named *)
let f2 ~named_module:{M : S} = body2

(** can be applied implicitly or explicitly *)
let f3 ?implicit_module:{M : S} = body3
```

The idea being : if a module argument has an implicit label and is omitted we use the implicit inference mechanism to infer the module expression.

Pros:
- A compact syntax for modular explicits and implicits.
- We are able to easily switch between implicit and explicit module arguments.
- If we have 3 implicits argument, and we want to instantiate the third by hand, we can by giving its label.

Cons:
- First class modules still cannot be used to write module dependent functions.
- We reuse the `?` symbol with a slightly different semantic (but we could choose a different symbol).
- All implicit arguments need 2 names (module name + label name) because both use a different set of namespace (but maybe `~M`, `?M` could be allowed?).

### 3 - Removing the split: unioning first-class modules and dependent module arguments

The idea of this syntax would be to allow functions taking an argument that is an unpacking of a first-class module to be dependent.

For example :

```ocaml
module type Show = sig
    type t
    val show : t -> string
end

let show_explicit (module S : Show) x = S.show x
(* val show_explicit : (module S : Show) -> S.t -> string *)

let () = print_string (show (module Show_Int) 3)
```

For implicit abstractions we could use the `implicit` keyword:

```
let show (implicit S : Show) x = S.show x
(* val show : (implicit S : Show) -> S.t -> string *)

let () =
    print_string (show 3);
    print_string (show (implicit Variant_Int_Show) 3)
```

Pros :
- Users will not have to choose if their functions use first-class modules or modular explicits as it is the same syntax for both.
- We can use the label mechanism with modular explicits (same as the syntax above)

Cons :
- Modular explicit and implicits becomes as verbose as first-class modules (use of the `module`/`implicit` keyword everywhere)

### 4 - New proposal : Both (2 + 3)

The idea here would be to have a mix of syntax 2 and 3 presented above. To put it simply all the examples above up to the `implicit` keyword would be accepted :
- functions taking a first-class module as argument can be made dependent,
- and modular explicits can be written with braces (`{M : S}`) (with labels for implicits).

```ocaml
module type Show = sig
    type t
    val show : t -> string
end

(** We can write module dependent functions with both syntax *)

let show_explicit (module S : Show) x = S.show x
(* val show_explicit : (module S : Show) -> S.t -> string *)

let show_explicit2 {S : Show} x = S.show x
(* val show_explicit2 : {S : Show} -> S.t -> string *)

(** Coercion is possible between them *)
let show_explicit3 = (show_explicit2 :> (module S : Show) -> S.t -> string)

(* Modular implicits use optional labels *)
let show {S : Show} x = S.show x

```

This would lead to 2 different incompatible syntax for passing module arguments to module-dependent functions. However, one would be strictly less expressive than the other but also the less verbose.

Pros :
- First-class modules can be dependent, and we have a compact syntax available for modular implicits,
- We are able to easily rewrite a function between implicit and explicit module arguments
- If we have 3 implicits argument, and we want to instantiate the third by hand, we can by giving its label.

Cons :
- We reuse the `?` symbol with a slightly different semantic (but we could choose a different symbol)
- All implicit arguments need 2 names (module name + label name) because both use a different set of namespace (but could be changed).

## Other possible extensions

When discussing syntax of modular explicits and modular implicits, syntax of other features could also be brought to your attention.

### (Semi-)Explicit type arguments

It could be useful for readability to be able to give a type argument to a function. Using modular explicits we will be able to give a module containing exactly a type to a dependent function but not directly the type.

A way to solve this could be to reuse the syntax for modular implicits presented above with braces or labels + braces and use this for type arguments :

```ocaml
let id {type a} (x : a) = x
(* val id : {type a} -> a -> a*)

let result (id : {type a} -> a -> a) = (id 3, id {type bool} true)
```

Such a mechanism would be a way to receive polymorphic arguments when writing a function.

We could also offer this with the `implicit` keyword or with labelled arguments:

```
let id (implicit type a) (x : a) = x
(* val id : (implicit type a) -> a -> a*)

let id ?a:(type a) (x : a) = x
(* val id : ?a:(type a) -> a -> a *)
```

### Signatures parametrized over a type

With modular implicits/explicits we could need to write a function demanding multiple modules with the same type. 

```ocaml
let compute {E : Eq}
    {A : Add with type t = E.t}
    {S : Sub with type t = E.t}
    {D : Default with type t = E.t} () = ...
```

However, this can become pretty verbose. A fix for this would be to allow parametric signatures because we know that those signatures will (almost) always be used with a `with type t = ...`.

This would allow us to write a much more compact code.

```ocaml
let compute {E : Eq} {A : Add(E.t)} {S : Sub(E.t)} {D : Default(E.t)} () = ...
```

However, this poses a few questions :
- Should we write `Add(E.t)` or maybe `Add(type E.t)` or `E.t Add` ?
- How should we write the definition of a parametric signature ?
`module Eq (type t) = type val eq : t -> t -> bool end`?
`module type t Eq = type val eq : t -> t -> bool end`?
- Should we always write the parameter of the parametric signature ? Or can it be inferred or abstracted when not written ?

**Comment by kit-ty-kate** [comment]
I personally think it's more natural to have:
- `(module M : S)` to denote modular explicits
- `{module M : S}` to denote modular implicits

The modular explicit syntax is very natural because `(module M : S)` is already accepted in the definitions, so allowing it in the types makes perfect sense.

And the modular implicit syntax is natural for both anyone who knows a bit of Coq (see [manual](https://coq.inria.fr/doc/v8.12/refman/language/extensions/implicit-arguments.html)) and current OCaml users who might be familiar with first-class modules. Having the `module` keyword there is also a nice addition to me as it allows an easier extension of the syntax in the future, as well as a more explicit descriptions of the different values being introduced (I don't think the `{M : S}` syntax makes it easy for beginner to understand what those values are (are they variants? modules?), whereas with the `module` keyword it makes it much more explicit.

**Reply by gasche** [reply]
Labelled arguments (optional or not) let us specify an argument by name, and I think this is very important in practice in languages that have implicitly-inferred arguments. For example in Coq, if you do `Definition Foo {A : Type} {Dec : decidable A} (x y : A) = ...`, the parameters `A` and `Dec` are implicit, but if you want you can call `Foo (Dec := ...)` to explicitly pass a `Dec` parameter.

We could consider adding exactly this syntax in OCaml, or a variant that avoids ambiguities. But this introduces the idea that non-labelled functions/abstractions can be applied by formal parameter name, that does not currently exist in the language. (It would also be useful for `(type a)` abstractions for example.)

On the other hand, we already have a mechanism for optional named parameters in OCaml, `?foo` or `?(foo : type)`, so `?(module M : S)` or `?(M : S)` is a very plausible extension for implicit parameters that can be explicitly applied by name. (With a different semantics for what to do when the argument is not passed explicitly, compared to normal optional arguments.)

TL;DR: maybe we don't need an optional-argument syntax, but we do need the possibility the pass these arguments explicitly by name.

**Reply by lpw25** [reply]
I don't think labelled arguments are that important for parameters filled in by search rather than by unification. Haskell doesn't even let you specify these arguments manually, and I'm not sure Agda lets you fill in instance arguments by name. I think it's reasonable to support `~foo:{Bar : S}`, which can be applied implicitly or explicitly with a label, but I don't think there is any need to arbitrarily conflate optional arguments with implicit arguments.

**Reply by gasche** [reply]
If you have `f : {{x : A}} -> B` in Agda and `a : A`, then you can use `f {{a}}` and `f {{x = a}}` to instantiate the instance argument by position or by name. (Scala implicits can be applied explicitly, but only by position I believe.)

I think that this matters more for designs where instances are not necessarily coherent. As long as you say either "it's okay if the search has two solutions, we'll just pick one" or, on the opposite end, "we fail in case of ambiguity, and you have to tell us which you mean", in either cases you want to be able to explicitly disambiguate by forcing a specific solution. Haskell was mostly design with coherence-by-construction in mind, which de-emphasizes this situation. I don't have enough experience with modular implicits to tell which way the design pendulum will go -- maybe that's what explains your own intuition.

**Comment by lpw25** [comment]
Personally, I prefer designs which keep modular implicits and dependent module parameters separate, and treat dependent module parameters and first-class module parameters as closely related.

One reason for this is that I think they will mostly be used for very different use cases, and be thought about by users very differently, with implicits thoughts about more like type-classes and dependent module parameters thoughts about more like ordinary parameters.

Another reason is that I think it is worth exploring the design space where the thing you write after the colon in an implicit parameter is not an ordinary module type. I'm particularly interested in exploring designs that look something like:
```ocaml
module type Set = sig
  type t
  type elt
  val empty : t
  val add : t -> elt -> t
  [etc]
end

type class Set[type t] : Set with type elt = t
instance Set[string] := String.Set
instance Set[bool] := Bit_set

let singleton (type a) {S : Set[a]} s = S.add s S.empty
```
which gives `singleton` a type like:
```ocaml
val singleton : {S : Set['a]} -> 'a -> S.t
```
I like this kind of design for a couple of reasons:
1. The user can control what the search is keyed on. Here we can only choose the `Set` based on the type of the element not the type of the set itself. I suspect this may also make the compile-time performance of search easier to think about.
2. It enables doing an initial version without needing to define unification on module expressions, where searches can only be keyed on types. It could be extended later to support searching by full module expressions.

I'm not saying we should definitely go in this direction, but I do think it is worth exploring, and it certainly pushes in the direction of thinking of implicits and dependent module parameters as quite different.

**Reply by samsa1** [reply]
Such a design with restriction to instances could destroy some use case of modular implicits that could have been possible using structural subtyping of modules.

For example if we take [ocamlgraph](https://github.com/backtracking/ocamlgraph) there is lost of cases of functions using different definitions of graphs with various properties. That library could use modular implicits by having every function defining the notion of graph it needs and definitions of graph could be as general as possible.
However, if we use instances we would need to give a way to go from bigger definitions to smaller when we require a less restrictive notion of graph than what was defined previously.

**Reply by lpw25** [reply]
It's true that it intentionally pushes in a more nominal direction, but it doesn't prevent you from adding structural features later if you want. For example, you could add support for writing something like this:
```ocaml
module type GraphS = ...

type class Graph[type t][module type S <: GraphS with type elt = t] : S
instance (type t) (module type S <: GraphS with type elt = t) (module type T <: S) {M : Graph[t][T]} 
  Graph[t][S] := M

module type Simple_graphS = ...
module type Graph_with_more_stuffS = ...

instance Graph[int][Simple_graphS] := A_simple_int_graph
instance Graph[float][Graph_with_more_stuffS] := A_fancy_float_graph

val empty : {G : Graph['a][GraphS]} -> unit -> G.t
```

The gain here is that people only get those more complicated behaviours when they ask for them, and we don't need to support them at all in version 1.

You can also leave out the structural features and just write out a nominal hierarchy of all the types in ocamlgraph. So:
```ocaml
type class Graph[type t] : GraphS with type t = t
type class Simple_graph[type t] : Simple_graphS with type t = t
type class Graph_with_more_stuff[type t] : Graph_with_more_stuffS with type t = t

instance (type t) {M : Graph[t]} Simple_graph[t]  := M
instance (type t) {M : Simple_graph[t]} := Graph_with_more_stuff[t] := M
```
This has the advantage of making explicit the intended hierarchy.

**Comment by Octachron** [comment]
A variation on 2, could we reuse the Coq-like grouping of similar argument like we do for locally abstract types
```ocaml
let f (type a b c) = ()
```
to factorize the syntactic cost:
```ocaml
let f (module (L:Label) (E: Edge with Label = L) (N:Node with E = Edge)) = ()
```
